### PR TITLE
[Inductor][CPP] Fix symbolic size issue of gemm template

### PR DESCRIPTION
**Pull request by leslie-fang-intel**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #131968

**Summary**
Fix https://github.com/pytorch/pytorch/issues/131929, after turning on `inline_inbuilt_nn_modules`, in the dynamic shape testing, the size of activation has changed from `[s0, 128]` to `[s0, s1]` and fails the check that https://github.com/pytorch/pytorch/blob/8b04edcac1f00a8c292a2de18dabd1138f7c3c99/torch/_inductor/utils.py#L1172

In this PR, when we found one of size K in activation or weight is constant, we will use this constant number instead of another symbolic value.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Comment by leslie-fang-intel** [comment]
Hi @anijain2305, thanks for the suggestion in https://github.com/pytorch/pytorch/issues/131929 regarding to fix this issue. I am trying to implement it in this PR.

However, after a second thought, I am wondering the behavior that we generate the size of activation as `[s0, s1]` is correct or not. Since it's the linear case, `s1` of activation should equals to the k dimension of weight which is a constant. Shouldn't dynamo specialize the `s1` of activation to the same constant?

**Review by jgong5 — CHANGES_REQUESTED** [review]
I wonder why k1 is symbolic while k2 is static with module inlining while both of them are static without. Is there any issue in the upper stack of the code?

**Comment by leslie-fang-intel** [comment]
This issue doesn't exist for bf16 and only happens after export/applying the quantization flow.
Here is a script to reproduce: https://gist.github.com/leslie-fang-intel/0a40db6fb6dec291fe6c73a7829e5c17

- BF16 Graph, the K dim is a constant:
```
V0728 21:49:44.999000 1517936 torch/_inductor/freezing.py:117] [0/0]  /localdisk/leslie/torch_inductor_community/pytorch/torch/fx/_lazy_graph_module.py class <lambda>(torch.nn.Module):
V0728 21:49:44.999000 1517936 torch/_inductor/freezing.py:117] [0/0]     def forward(self, arg2_1: "Sym(s0)", arg3_1: "f32[s0, 128]"):
V0728 21:49:44.999000 1517936 torch/_inductor/freezing.py:117] [0/0]          # File: /localdisk/leslie/torch_inductor_community/torch_script/inductor/symbolic_size/test_linear.py:28 in forward, code: res = self.epilogue(self.linear(x))
V0728 21:49:44.999000 1517936 torch/_inductor/freezing.py:117] [0/0]         _frozen_param2: "f32[128, 64]" = self._frozen_param2
V0728 21:49:44.999000 1517936 torch/_inductor/freezing.py:117] [0/0]         mm: "f32[s0, 64]" = torch.ops.aten.mm.default(arg3_1, _frozen_param2);  arg3_1 = _frozen_param2 = None
V0728 21:49:44.999000 1517936 torch/_inductor/freezing.py:117] [0/0]         relu: "f32[s0, 64]" = torch.ops.aten.relu.default(mm);  mm = None
V0728 21:49:44.999000 1517936 torch/_inductor/freezing.py:117] [0/0]         
V0728 21:49:44.999000 1517936 torch/_inductor/freezing.py:117] [0/0]          # File: /localdisk/leslie/torch_inductor_community/torch_script/inductor/symbolic_size/test_linear.py:29 in forward, code: res = self.epilogue2(self.linear2(res))
V0728 21:49:44.999000 1517936 torch/_inductor/freezing.py:117] [0/0]         _frozen_param3: "f32[64, 64]" = self._frozen_param3
V0728 21:49:44.999000 1517936 torch/_inductor/freezing.py:117] [0/0]         mm_1: "f32[s0, 64]" = torch.ops.aten.mm.default(relu, _frozen_param3);  relu = _frozen_param3 = None
V0728 21:49:44.999000 1517936 torch/_inductor/freezing.py:117] [0/0]         relu_1: "f32[s0, 64]" = torch.ops.aten.relu.default(mm_1);  mm_1 = None
V0728 21:49:44.999000 1517936 torch/_inductor/freezing.py:117] [0/0]         return (relu_1,)
```

- INT8 Graph, the K dim has changed to be symbolic
```
V0728 22:04:32.979000 1522857 torch/_inductor/freezing.py:117] [1/0]     def forward(self, arg6_1: "Sym(s4)", arg7_1: "Sym(s5)", arg8_1: "f32[s4, s5]"):
V0728 22:04:32.979000 1522857 torch/_inductor/freezing.py:117] [1/0]         # No stacktrace found for following nodes
V0728 22:04:32.979000 1522857 torch/_inductor/freezing.py:117] [1/0]         _frozen_param0: "f32[s10]" = self._frozen_param0
V0728 22:04:32.979000 1522857 torch/_inductor/freezing.py:117] [1/0]         _frozen_param1: "i64[s10]" = self._frozen_param1
V0728 22:04:32.979000 1522857 torch/_inductor/freezing.py:117] [1/0]         _frozen_param3: "f32[s10]" = self._frozen_param3
V0728 22:04:32.979000 1522857 torch/_inductor/freezing.py:117] [1/0]         _frozen_param4: "i64[s10]" = self._frozen_param4
V0728 22:04:32.979000 1522857 torch/_inductor/freezing.py:117] [1/0]         
V0728 22:04:32.979000 1522857 torch/_inductor/freezing.py:117] [1/0]          # File: <eval_with_key>.12:7 in forward, code: quantize_per_tensor_default = torch.ops.quantized_decomposed.quantize_per_tensor.default(arg0_1, 0.021388564258813858, 124, 0, 255, torch.uint8);  arg0_1 = None
V0728 22:04:32.979000 1522857 torch/_inductor/freezing.py:117] [1/0]         quantize_per_tensor: "u8[s4, s5]" = torch.ops.quantized_decomposed.quantize_per_tensor.default(arg8_1, 0.021388564258813858, 124, 0, 255, torch.uint8);  arg8_1 = None
V0728 22:04:32.979000 1522857 torch/_inductor/freezing.py:117] [1/0]         
V0728 22:04:32.979000 1522857 torch/_inductor/freezing.py:117] [1/0]         # No stacktrace found for following nodes
V0728 22:04:32.979000 1522857 torch/_inductor/freezing.py:117] [1/0]         _frozen_param6 = self._frozen_param6
V0728 22:04:32.979000 1522857 torch/_inductor/freezing.py:117] [1/0]         
V0728 22:04:32.979000 1522857 torch/_inductor/freezing.py:117] [1/0]          # File: /localdisk/leslie/torch_inductor_community/torch_script/inductor/symbolic_size/test_linear.py:34 in forward, code: res = self.epilogue(self.linear(x))
V0728 22:04:32.979000 1522857 torch/_inductor/freezing.py:117] [1/0]         qlinear_pointwise_default_1: "f32[s4, s10]" = torch.ops.onednn.qlinear_pointwise.default(quantize_per_tensor, 0.021388564258813858, 124, _frozen_param6, _frozen_param0, _frozen_param1, None, 1.0, 0, torch.float32, 'none', [], '');  quantize_per_tensor = _frozen_param6 = _frozen_param0 = _frozen_param1 = None
V0728 22:04:32.979000 1522857 torch/_inductor/freezing.py:117] [1/0]         relu: "f32[s4, s10]" = torch.ops.aten.relu.default(qlinear_pointwise_default_1);  qlinear_pointwise_default_1 = None
V0728 22:04:32.979000 1522857 torch/_inductor/freezing.py:117] [1/0]         
V0728 22:04:32.979000 1522857 torch/_inductor/freezing.py:117] [1/0]          # File: <eval_with_key>.12:15 in forward, code: quantize_per_tensor_default_1 = torch.ops.quantized_decomposed.quantize_per_tensor.default(relu, 0.006081726402044296, 0, 0, 255, torch.uint8);  relu = None
V0728 22:04:32.979000 1522857 torch/_inductor/freezing.py:117] [1/0]         quantize_per_tensor_1: "u8[s4, s10]" = torch.ops.quantized_decomposed.quantize_per_tensor.default(relu, 0.006081726402044296, 0, 0, 255, torch.uint8);  relu = None
V0728 22:04:32.979000 1522857 torch/_inductor/freezing.py:117] [1/0]         
V0728 22:04:32.979000 1522857 torch/_inductor/freezing.py:117] [1/0]         # No stacktrace found for following nodes
V0728 22:04:32.979000 1522857 torch/_inductor/freezing.py:117] [1/0]         _frozen_param7 = self._frozen_param7
V0728 22:04:32.979000 1522857 torch/_inductor/freezing.py:117] [1/0]         
V0728 22:04:32.979000 1522857 torch/_inductor/freezing.py:117] [1/0]          # File: /localdisk/leslie/torch_inductor_community/torch_script/inductor/symbolic_size/test_linear.py:35 in forward, code: res = self.epilogue2(self.linear2(res))
V0728 22:04:32.979000 1522857 torch/_inductor/freezing.py:117] [1/0]         qlinear_pointwise_default: "f32[s4, s10]" = torch.ops.onednn.qlinear_pointwise.default(quantize_per_tensor_1, 0.006081726402044296, 0, _frozen_param7, _frozen_param3, _frozen_param4, None, 1.0, 0, torch.float32, 'none', [], '');  quantize_per_tensor_1 = _frozen_param7 = _frozen_param3 = _frozen_param4 = None
V0728 22:04:32.979000 1522857 torch/_inductor/freezing.py:117] [1/0]         relu_1: "f32[s4, s10]" = torch.ops.aten.relu.default(qlinear_pointwise_default);  qlinear_pointwise_default = None
V0728 22:04:32.979000 1522857 torch/_inductor/freezing.py:117] [1/0]         return (relu_1,)
```

**Comment by leslie-fang-intel** [comment]
Hi @anijain2305, it looks like a issue in Dynamo. Could you help to take a look? or do you have any suggestion for how to further debug?

**Comment by leslie-fang-intel** [comment]
Will fix in https://github.com/pytorch/pytorch/pull/132098 with upper stack of the code.

**Comment by leslie-fang-intel** [comment]
Will fix in https://github.com/pytorch/pytorch/pull/132098 with upper stack of the code.
